### DISCUSSION TITLE
A pool memory resource for usm

### DISCUSSION THREAD

**Discussion by RaulPPelaez** [AdaptiveCpp/AdaptiveCpp#972] [Q&A] [answered]
I am trying to reduce allocation/deallocation overhead by implementing a pool allocator. I am having trouble understanding how to uniquely identify a queue.

In CUDA, I would create a pool memory resource per device and draw memory from one or the other depending on cudaGetDevice. That is, each of the possible devices is uniquely identified by a hash returned by cudaGetDevice.

In SYCL, the underlying malloc function, say ```sycl::malloc_shared(std::size_t num_bytes, sycl::queue q)```, requires a queue.  Thus I cannot apply the same strategy here since queue carries much more context than a simple device ID (for instance, it might also control a cuda stream).
That makes it a bad idea to have a per-queue memory pool (and AFAIK, SYCL does not provide a way to uniquely hash a queue anyway).
I can think of, for instance, controlling queue creation/destruction by wrapping sycl::queue and keeping a set of them, but the standard does not seem to provide ways to even compare two queues.

It feels like I am trying to do something in a way that SYCL really does not want me to, so I wonder:

How would you approach the design of a pool memory resource in SYCL?

**Comment by illuhad** [comment, marked as answer]
I'm not sure I understand the issue. The `malloc_shared` signature you mention is not the only overload. There's also `malloc_shared(std::size_t num_bytes, const sycl::device& dev, const sycl::context& ctx)`.

The variant taking queue is just a shortcut for `malloc_shared(num_bytes, q.get_device(), q.get_context())`.
In principle you would need to pool based on `context` and `device`, however, in Open SYCL `context` is pretty meaningless anyway so you could just ignore it.

> but the standard does not seem to provide ways to even compare two queues.

This is not true. `queue` follows what SYCL calls "common-reference semantics", which among other things requires that comparison be supported.
https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:reference-semantics

**Comment by RaulPPelaez** [comment]
Thanks for the insights!

I missed the other versions of ```malloc_shared```, I believe I can leverage them for what I want to do. In particular I see that this passes

```c++
TEST(Queue, EquivalentDevicesAreEqual) {
  auto q1 = sycl::queue{sycl::default_selector{}};
  auto q2 = sycl::queue{sycl::default_selector{}};
  ASSERT_TRUE(q1.get_device() == q2.get_device());
}
```

On the other hand, I noticed this does not pass:

```c++

TEST(Queue, EquivalentContextsAreEqual) {
  auto q1 = sycl::queue{sycl::default_selector{}};
  auto q2 = sycl::queue{q1.get_device()};
  ASSERT_TRUE(q1.get_context() == q2.get_context());
}
```
(see that I am not creating the second queue using the context of the first). With what I gathered from the standard I would expect both contexts to be equal in this case. 

Thanks, this was super helpful!

**Reply by illuhad** [reply]
> With what I gathered from the standard I would expect both contexts to be equal in this case.

Unless explicitly given a context, each queue will by default construct its own context in the constructor. This is why they do not compare equal. This has caused a lot of headaches for a lot of people, especially on implementations such as DPC++ where the context actually matters and you cannot just access data allocated in a different context. As I said, in Open SYCL context is largely meaningless, so you can ignore this :-)
